src/Workspaces/Remote/ServiceHub/Services/CodeAnalysisService_CodeLens.cs

```diff
@@ -63,12 +87,155 @@ public Task<string> GetFullyQualifiedName(DocumentId documentId, TextSpan textSp
+            private readonly DocumentId _documentId;
+            private readonly CancellationToken _cancellationToken;
+
+            private bool _eventSubscribed;
+            private VersionStamp _lastVersion;
+            private ResettableDelay _resettableDelay;
+
+            public static async Task TrackAsync(CodeAnalysisService owner, Workspace workspace, DocumentId documentId, CancellationToken cancellationToken)
+            {
+                var document = workspace.CurrentSolution.GetDocument(documentId);
+                if (document == null)
+                {
+                    return;
+                }
+
+                // if anything under the project this file belong to changes, then invalidate the code lens so that it can refresh
+                var dependentVersion = await document.Project.GetDependentVersionAsync(cancellationToken).ConfigureAwait(false);
+                var _ = new WorkspaceChangeTracker(owner, workspace, documentId, dependentVersion, cancellationToken);
+            }
+
+            private WorkspaceChangeTracker(
+                CodeAnalysisService owner,
+                Workspace workspace,
+                DocumentId documentId,
+                VersionStamp dependentVersion,
+                CancellationToken cancellationToken)
+            {
+                _gate = new object();
+                _eventSubscribed = false;
+
+                _owner = owner;
+                _workspace = workspace;
+                _documentId = documentId;
+                _cancellationToken = cancellationToken;
+
+                _lastVersion = dependentVersion;
+                _resettableDelay = ResettableDelay.CompletedDelay;
+
+                ConnectEvents(subscription: true);
+            }
+
+            private void ConnectEvents(bool subscription)
+            {
+                // this is only place lock is used.
+                // we have a lock here so that subscription and unsubscription of the two events
+                // are happening atomic, but that doesn't mean there is no possiblity of race here
+                // theoradically, there can be a race if connection got disconnected before we subscribe
+                // to OnRpcDisconnected but already in the subscription code path.
+                // but there is no easy way to solve the problem unless Rpc itself provide things like
+                // subscribe only if connection still alive or something
+                lock (_gate)
+                {
+                    if (subscription)
+                    {
+                        _owner.Disconnected += OnDisconnected;
+                        _workspace.WorkspaceChanged += OnWorkspaceChanged;
+
+                        if (_cancellationToken.IsCancellationRequested)
+                        {
+                            // while, we are subscribing to this service, caller side closed this connection
+                            // unsubscribe from the service
```

Comment:
But the cancellation can still happen after the check, right?

Reply:
ya. but at that point, we are already subscribed so, the event handler will handle the cancellation (disconnect) situation.